Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm

```diff
@@ -723,14 +723,30 @@ static void registerLogHook()
         if (msg->buffer_sz > 1024)
             return;
 
-        // Skip faults and debug logs in non-internal builds.
-        if (type == OS_LOG_TYPE_FAULT || type == OS_LOG_TYPE_DEBUG)
+        // Skip debug logs in non-internal builds.
+        if (type == OS_LOG_TYPE_DEBUG)
             return;
 
         CString logFormat(msg->format);
         CString logChannel(msg->subsystem);
         CString logCategory(msg->category);
 
+        // Don't send fault logs on background queue, in case the WebContent process will be terminated.
+        // If the WebContent process is terminated, we might not be able to send the log in time.
+        if (type == OS_LOG_TYPE_FAULT) {
+            char* messageString = os_log_copy_message_string(msg);
+            if (!messageString)
+                return;
+            IPC::DataReference logString(reinterpret_cast<uint8_t*>(messageString), strlen(messageString) + 1);
+
+            auto connectionID = WebProcess::singleton().networkProcessConnectionID();
+            if (connectionID)
+                IPC::Connection::send(connectionID, Messages::NetworkConnectionToWebProcess::LogOnBehalfOfWebContent(logChannel.bytesInludingNullTerminator(), logCategory.bytesInludingNullTerminator(), logString, OS_LOG_TYPE_ERROR, getpid()), 0);
```

Comment:
I suggest with still use the dispatch queue but use WorkQueue::dispatchWithQoS() to make the message priority higher. Otherwise, we can end up with out-of-order logging which would be unfortunate.

Reply:
We should also pass the QoS to IPC::Connection::send() for FAULT messages (already supported).

Reply:
Ah, good point, will fix!

Thanks for reviewing!